Added tab selection ui component, with working example in iGPSe workflow

chaitanya2334 opened the pull request:
For now, the tabbed interface only takes in one image per tab. The UI spec has to be specified with the list along with what you would generally mention for an image type UI component. An example is present in iGPSe2. 

As of march 15th deadline, I think this should suffice.

But, please do let me know if this component needs to be more flexible, ( like multiple images per "tab-page", or it should allow multiple analysis elements per "tab-page")

Commit:
Added tab selection ui component, with working example in iGPSe workflow

Commit:
removed unused code and a few lint errors seen by travis

Commit:
fixed some more travis lint errors

opadron commented:
@chaitanya2334 thanks for putting this in.  I'll have a look at this later today.
Is this something you would be able to demo for us at the call?

Commit:
revamp analysis reducers/clean up tab component

As I was studying Chaitanya's changes, I was stuck
trying to figure out a solution for the element
update problem (the details of which wouldn't make
sense unless you were also familiar with the other
PRs that I will need to integrate with this one).
I ended up deciding that the fastest way to solve
the problem was to just implement the longer-term
feature of allowing analysis elements to be
arbitrarily nested.  Turns out it wasn't as much
work as I thought!  This give us the added benefit
of putting *anything* in those tabs, now -- not
just images!

 - renamed tab selection widget to "tab group"

   - filename is now
     "web_client/.../tab-group.jsx"

   - component class name is now "TabGroup"

   - analysis element type field value is now
     "tabGroup"

 - revamped analysis reducers

   - Analysis elements can now be parents of other
     analysis elements!  New element type
     implementations should consult their
     "children" prop and decide what, if anything,
     to do with them.  TabGroups can now hold any
     kind of analysis element, including other
     TabGroups!

   - Analysis element state attribute "forms" has
     been removed.  It is replaced with a new
     "states" attribute, which is a flat elementId
     -> elementState mapping, unlike the original
     forms that tried to heirarchicaly organize
     state data.

   - If an object representing a new element to be
     added contains an "elements" attribute with
     similarly constructed subelements, all such
     subelements will now be automatically added
     when adding the top-level element.  This
     change applies recursively to all added
     subelements.

   - If an object representing a new page to be
     added contains an "elements" attribute with
     objects representing new elements, all such
     elements will now be automatically added
     when adding the page.  Any such elements are
     added in the recursive fashion described
     above.

 - global action creator "triggerAnalysisAction"
   no longer takes the analysis form as its first
   argument.  Instead, it is replaced by an object
   with two attributes:

   - objects: copy of the analysis object cache

   - states: copy of the analysis state cache

 - analysisUtils.aggregateForm has been replaced
   by aggregateStateData, which better describes
   the utility function's purpose and does a
   better job using the new analysis state cache.

 - tweaked ui specs and analysis modules to take
   advantage of the new changes.

opadron commented:
To move the process along, I'm approving this PR.
See #42 for integration.
